### PR TITLE
Challenge mode

### LINKED ISSUE #15: Challenge mode
Implement a Challenge mode where users try to complete preset challenges.  A challenge consists of 

* Some kind of description of a finite world (it isn't yet clear to me what the best way is to specify the world)
* An initial inventory for the base
* An unrestricted global program (of type `cmd bool`) which runs to completion on every tick to see whether the challenge has been successfully completed.

It should be possible to describe challenges in a `.yaml` file of some to-be-determined format.

### PR DESCRIPTION
**Pull request by byorgey**
Add a "challenge mode" where the player tries to achieve some specified goal from specific starting conditions.

This is very much a work in progress, but feedback could be useful already.  Still to do:

- [x] Add inventory parsing for robots
- [x] Allow defining custom entities?
- [X] Add parsing for worlds
- [X] Make a way for the user to choose a challenge (perhaps just via command-line flag until we implement #4 ).
- [ ] Add some actual challenges!

Closes #15 .

### CODE REVIEW

**Comment by xsebek** [comment]
This is shaping up great! :star_struck: 

Some thoughts:
- I really like that I can continue playing past the win
- when `Won` is first set, it should be displayed somewhere, maybe reuse the Error window?
- every base should have a `logger` otherwise it is really confusing
- there should be a challenge description sent to either a logger (easy) or a separate <kbd>F2</kbd> window with checkmarks (cool), IMHO both would be best :slightly_smiling_face: 
- the challenge file should be searched with literal path too (and later maybe relative to some XDG directory) 
- the blank file could be nice for mazes, maybe add a Filled file too? :eyes: 
- the God capability and `as` is pretty cool :innocent: 
  - could robots "get" a God item and lose it when switching modes? It could show which God commands are available
- could you make the current Creative and Classic modes Challenges too?
  - we could specify name for challenge ("Creative mode"/"Move tutorial"/...)
  - that way #50 would be half done by reusing challenge files :open_mouth: (just save them in XDG instead)
- once you finish this, we could start versioning and releasing the game, as challenges are pretty cool :thinking: 

I hope this is not too much, thanks for working on this @byorgey :+1:

**Comment by xsebek** [comment]
Btw. could the `as` command run in another thread? Long computation freeze the UI and can not be stopped. :confused: 

I guess the situation is not super likely so I am fine leaving it to another PR (and I would be happy to try [`async`](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#v:withAsync)) :wink:

**Comment by byorgey** [comment]
Thanks for the feedback @xsebek !

> I really like that I can continue playing past the win

Yep, that's definitely intentional.

> when Won is first set, it should be displayed somewhere, maybe reuse the Error window?

Right now for testing I just had it change the display of the mode from `Challenge` to `Challenge!` but having a window pop up is definitely the plan.  I'll probably do that soon.

> every base should have a logger otherwise it is really confusing

Good idea, though I think this is something we should just leave up to challenge writers (we could even have some documentation on best practices for creating challenges) rather than having it specially built in.

> there should be a challenge description sent to either a logger (easy) or a separate F2 window with checkmarks (cool), IMHO both would be best

Yeah, I was thinking of having some kind of special item whose description tells you about the challenge.  Having a built-in key like F2 for making it pop up sounds like a good idea too.  I'm not sure what you mean by checkmarks?

> the challenge file should be searched with literal path too (and later maybe relative to some XDG directory)

Sure, good idea.

> the blank file could be nice for mazes, maybe add a Filled file too?

You can already specify the default terrain and entity to be used for any cells that aren't explicitly specified. So e.g. you could have a water world or whatever. Blank is just what you get if you don't specify a default (i.e. it's the "default default" :wink: ).   I also intend to have a mode where you specify a seed, and any cells not explicitly specified in the map will be filled in using the seed just like they would be in classic mode.

> could robots "get" a God item and lose it when switching modes? It could show which God commands are available

I was mostly intending the God commands to be used for authoring challenges, but I suppose they could be useful in creative mode as well.  I'm not sure I like the idea of items appearing and disappearing from the inventory; also, the ability to switch between modes is really only for testing and will probably go away at some point.  But maybe when you start a game in Creative mode you start with a special item like a "wand of power" or something like that which tells you about available commands?

> could you make the current Creative and Classic modes Challenges too?

Ah, hmm, I see... so instead of just "challenges" they would be something a bit more general, just describing how to initialize a game.  I think that sounds like a good idea!

> that way Saving/loading games #50 would be half done by reusing challenge files open_mouth (just save them in XDG instead)

Hmm, maybe, but there's a lot more information you need in a save file than in a challenge description...

**Comment by byorgey** [comment]
> Btw. could the `as` command run in another thread? Long computation freeze the UI and can not be stopped.

Yes, that's definitely a danger.  This is a consequence of the need to have win condition checking run to completion every tick, without the usual constraints on the number of CESK machine steps.  I don't think we can really run it in another thread because we need to know right away whether you have completed a challenge or not.  The code for checking the win condition is supposed to run every tick; running it asynchronously isn't a good idea since it could just get farther and farther behind the main game loop.

Again, if the `as` command is only used in designing challenges then I'm not too worried; this is just something you have to think about carefully when designing and testing your challenge.  If it is also available in creative mode then it would be possible to freeze your game accidentally, which would be unfortunate.  Perhaps this is an argument in favor of having certain "dangerous" commands only available to "system" robots, and *not* available to players in creative mode.

**Comment by xsebek** [comment]
> "dangerous" commands only available to "system" robots, and not available to players in creative mode.

I would not worry about it, if some Player/God is testing a challenge in creative mode it would be annoying if we made it harder.

> I'm not sure what you mean by checkmarks?

I was thinking ahead (so no need to do it here) about a multistage challenge like the tutorial - find tree, bring it back, make a log,... 

> like a "wand of power" or something like that which tells you about available commands?

My idea was that you would not need to have the item in inventory, it would be shown in the list of robot with the game mode set to Creative. It would only look like it is on inventory. Similar trick could be used to add the challenge description.

**Comment by byorgey** [comment]
> there should be a challenge description sent to either a logger (easy) or a separate F2 window with checkmarks (cool), IMHO both would be best

For now I just put a special custom "challenge" entity in the inventory.  We could certainly add some code to look for this and handle it in some special way, I am not sure exactly how.

**Comment by byorgey** [comment]
I am pretty happy with this so far.  I think it is certainly ready to have people play with it and try building their own challenges.  There are a bunch more things I can think of to improve it; I am wondering what else is critical to do as part of this PR, and what we can leave to future improvement.  Off the top of my head:

- [ ] Improve the challenge file search mechanism (cf @xsebek 's comment "the challenge file should be searched with literal path too (and later maybe relative to some XDG directory)")
- [ ] Provide a mechanism for a sequence of challenges, e.g. a tutorial where after doing one challenge you can progress to the next.  Perhaps put a bunch of challenges in a list in a single file, and provide a way for the user to signal when they are ready to progress to the next challenge.
- [ ] Perhaps some improvements to the way challenge descriptions/goals are shown, winning is flagged, etc.
- [ ] Ability to specify a world seed instead of a default cell content, so we can have challenges that take place in random worlds a la classic mode
- [ ] Possibly generalize challenges to just descriptions of how to initialize a game --- so classic and creative modes can be "just" generated by descriptions of how to initialize the game rather than special built-in modes.

**Review by xsebek — CHANGES_REQUESTED** [review]
The challenge has a "codex" in robot inventory, but once I change that, it works nicely. :slightly_smiling_face: 

There are two things I would like to be done in this PR:
- a unit test for parsing the files in `data/challenges`
- writing a relative or absolute file path on command line should look in that path
  - we do not give hints which challenges are available, so it makes sense to use shell <kbd>Tab</kbd> to find the path
  - as a matter of principle I do not like requiring the challenges to be in one specific path when there is no technical reason for it

I hope my last point does not come across too strongly, but I imagine challenges will be written *by players for players* too and forking a repo seems unnecessary. :sweat_smile:

File: src/Swarm/Language/Pipeline.hs [hunk 1]
```diff
@@ -52,7 +53,15 @@ data ProcessedTerm
...
       -- ^ Capability context for any definitions embedded in the term
-  deriving (Data)
+  deriving (Data, Show)
+
+instance FromJSON ProcessedTerm where
+  parseJSON = withText "Term" tryProcess
+   where
+    tryProcess :: Text -> Y.Parser ProcessedTerm
+    tryProcess t = case processTerm t of
+      Left err -> fail $ "Error while processing win condition: " ++ from err
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Language/Pipeline.hs]: Right now it is only parsed in win condition, but you're right!  We might reuse it elsewhere later.

File: src/Swarm/Game/State.hs [hunk 2]
```diff
@@ -123,6 +134,19 @@ data REPLStatus
     REPLWorking Polytype (Maybe Value)
   deriving (Eq, Show)
 
+data WinCondition
+  = -- | There is no winning condition (e.g. we are in Classic or
+    --   Creative mode, or the player has already won and been
+    --   congratulated).
+    NoWinCondition
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/Game/State.hs:140]: Sure, good idea.  Done.

File: src/Swarm/TUI/View.hs [hunk 3]
```diff
@@ -235,24 +231,31 @@ drawDialog s = case s ^. uiModal of
...
   gameModeWidget =
     padLeft Max . padLeftRight 1
       . txt
       . (<> " mode")
       $ case mode of
-        Classic -> "Classic"
-        Creative -> "Creative"
+        ClassicMode -> "Classic"
+        CreativeMode -> "Creative"
+        ChallengeMode -> "Challenge"
```
[the remark this answers is not in the document]
Comment (byorgey) [reply src/Swarm/TUI/View.hs:258]: Yeah, the `M` key just cycles between the available modes.  I think you're right there is something strange going on here.

How about this: there's just a `Boolean` telling you whether you're in creative/God mode or not.  For now we can leave the `M` key to toggle this.  I had been thinking of challenges as a different "mode" but it's really not.  It's just a different way to initialize a world (and remember the fact that there's a win condition).

**Review by xsebek — APPROVED** [review]
LGTM! :+1: I checked it out and it all just works :slightly_smiling_face:

**Pull request merged**